File: java/org.financer.client/src/test/java/org/financer/client/connection/ServerRequestTest.java
package org.financer.client.connection;

import okhttp3.HttpUrl;
import okhttp3.mockwebserver.Dispatcher;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.financer.client.domain.model.user.User;
import org.financer.client.local.Application;
import org.financer.client.local.LocalStorage;
import org.financer.shared.domain.model.api.user.UserDTO;
import org.financer.shared.domain.model.value.objects.Email;
import org.financer.shared.domain.model.value.objects.Name;
import org.financer.shared.path.PathBuilder;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.Collections;
import java.util.Objects;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@Tag("unit")
public class ServerRequestTest {

    private final MockWebServer server = new MockWebServer();
    private HttpUrl httpUrl;

    @BeforeEach
    public void setUp() throws IOException {
        if (ServerRequest.MEDIA_TYPE_JSON == null) {
            throw new IllegalArgumentException("Illegal media type");
        }

        final LocalStorage localStorage = Mockito.mock(LocalStorage.class);
        when(localStorage.readObject(any(String.class))).thenReturn(null);

        ServerRequest.setLocalStorage(localStorage);
        ServerRequest.setApplication(Mockito.mock(Application.class));

        server.setDispatcher(new Dispatcher() {
            @NotNull
            @Override
            public MockResponse dispatch(@NotNull RecordedRequest request) {
                switch (Objects.requireNonNull(request.getPath())) {
                    case "/api/users":
                        return new MockResponse()
                                .setResponseCode(200)
                                .setHeader("Content-Type", ServerRequest.MEDIA_TYPE_JSON)
                                .setBody(new UserDTO()
                                        .setEmail(new Email("test@test.com"))
                                        .setName(new Name("test", "Name"))
                                        .setId(1)
                                        .toJson());
                    case "/api/categories":
                        return new MockResponse()
                                .setResponseCode(403)
                                .setBody(new RestErrorMessage()
                                        .setStatusCode(403)
                                        .setError("Unauthorized")
                                        .setTimestamp(LocalDateTime.now())
                                        .setMessages(Collections.singletonList("The user is not allowed to perform this operation."))
                                        .toJson());

                    default:
                        return new MockResponse().setResponseCode(404);
                }
            }
        });

        // Start the server.
        server.start();
        httpUrl = server.url("/api");

        when(localStorage.readObject("host")).thenReturn(httpUrl.toString());
        when(localStorage.contains("host")).thenReturn(true);
    }

    @Test
    public void testCreateServerRequestSuccess() {
        ServerRequest<User> serverRequest = new ServerRequest<>(new RequestConfig(
               PathBuilder.Get().users().build()), User.class, result -> {
            assertThat(result).isNotNull();
            assertThat(result.getEmail()).isEqualTo(new Email("test@test.com"));
        });
        serverRequest.run();
    }

    @Test
    public void testCreateServerRequestUnauthorized() {
        ServerRequest<User> serverRequest = new ServerRequest<>(new RequestConfig(
                PathBuilder.Get().categories().build()), User.class, new RestCallback<>() {
            @Override
            public void onSuccess(User result) {

            }

            @Override
            public void onFailure(Exception exception) {
                assertThat(exception).isNotNull();
            }
        });
        serverRequest.run();
    }

    @AfterEach
    public void tearDown() throws IOException {
        server.shutdown();
    }

}

File: java/org.financer.client/src/test/kotlin/org/financer/client/connection/ServerRequestTest.kt
package org.financer.client.connection

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import okhttp3.HttpUrl
import okhttp3.mockwebserver.Dispatcher
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okhttp3.mockwebserver.RecordedRequest
import org.financer.client.connection.ServerRequest.Companion.setLocalStorage
import org.financer.client.domain.model.user.User
import org.financer.client.local.Application
import org.financer.client.local.LocalStorage
import org.financer.shared.domain.model.api.user.UserDTO
import org.financer.shared.domain.model.value.objects.Email
import org.financer.shared.domain.model.value.objects.Name
import org.financer.shared.path.PathBuilder
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import java.io.IOException
import java.time.LocalDateTime
import java.util.*

@Tag("skip")
@ExperimentalCoroutinesApi
class ServerRequestTest {

    private val server = MockWebServer()
    private var httpUrl: HttpUrl? = null

    @BeforeEach
    @Throws(IOException::class)
    fun setUp() {
        val localStorage = Mockito.mock(LocalStorage::class.java)
        Mockito.`when`<Any?>(localStorage.readObject(ArgumentMatchers.any(String::class.java))).thenReturn(null)
        setLocalStorage(localStorage)
        ServerRequest.application = Mockito.mock(Application::class.java)
        server.dispatcher = object : Dispatcher() {
            override fun dispatch(request: RecordedRequest): MockResponse {
                return when (Objects.requireNonNull(request.path)) {
                    "/api/users" -> MockResponse()
                            .setResponseCode(200)
                            .setHeader("Content-Type", ServerRequest.MEDIA_TYPE_JSON)
                            .setBody(UserDTO()
                                    .setEmail(Email("test@test.com"))
                                    .setName(Name("test", "Name"))
                                    .setId(1)
                                    .toJson())
                    "/api/categories" -> MockResponse()
                            .setResponseCode(403)
                            .setBody(RestErrorMessage()
                                    .setStatusCode(403)
                                    .setError("Unauthorized")
                                    .setTimestamp(LocalDateTime.now())
                                    .setMessages(listOf("The user is not allowed to perform this operation."))
                                    .toJson())
                    else -> MockResponse().setResponseCode(404)
                }
            }
        }

        // Start the server.
        server.start()
        httpUrl = server.url("/api")
        Mockito.`when`<Any?>(localStorage.readObject("host")).thenReturn(httpUrl.toString())
        Mockito.`when`(localStorage.contains("host")).thenReturn(true)
    }


    @Test
    fun testCreateServerRequestSuccess() = runBlockingTest {
        val result = ServerRequest(RequestConfig(PathBuilder.Get().users().build())).execute<User?>()
        assert(result != null)
        assert(result!!.email!! == Email("test@test.com"))
    }

    @Test
    fun testCreateServerRequestUnauthorized() = runBlockingTest {
        val result = ServerRequest(RequestConfig(PathBuilder.Get().categories().build())).execute<User?>()
        assert(result != null)
    }

    @AfterEach
    @Throws(IOException::class)
    fun tearDown() {
        server.shutdown()
    }
}